compiler/semexprs.nim

```diff
@@ -1513,6 +1513,9 @@ proc semSubscript(c: PContext, n: PNode, flags: TExprFlags): PNode =
       result = n
     else:
       result = nil
+  of tyGenericParam:
+    result = n
+    result.typ = makeTypeFromExpr(c, n.copyTree)
```

Comment:
That seems to be a bad idea, can't we add a better proc/syntax/interface to ``tyGenericParam``. The ``[]`` is very overloaded already, seems to be recipe for desaster.

Reply:
Not sure I understand. If `[]` was a proc defined in system then this change wouldn't be required, but nkBracketExpr is still the special beast that is handled by its own rules. These rules don't handle `tyGenericParam`, hence the fix.

Another solution would be to define all missing procs in system.nim
```nim
proc `[]`[T: tuple](x: T, i: int): T {.magic: "XXX".}
```
Possibly others and then remove most of `semSubscript` and rely on operator overloading. 
Not sure this change will be easy and go smoothly though.

Reply:
`[]` special rules indeed cause a lot of issues (eg https://github.com/nim-lang/Nim/issues/15911) and should be replaced by magics